src/main/java/org/springframework/data/util/NullableWrapperConverters.java

```diff
@@ -0,0 +1,579 @@
+			super(io.vavr.control.Option.none(), Collections.singleton(io.vavr.control.Option.class));
+		}
+
+		public static WrapperType getWrapperType() {
+			return WrapperType.singleValue(io.vavr.control.Option.class);
+		}
+
+		/*
+		 * (non-Javadoc)
+		 * @see org.springframework.data.repository.util.QueryExecutionConverters.AbstractWrapperTypeConverter#wrap(java.lang.Object)
+		 */
+		@Override
+		protected Object wrap(Object source) {
+			return io.vavr.control.Option.of(source);
+		}
+	}
+
+	/**
+	 * A {@link Converter} to unwrap Guava {@link Optional} instances.
+	 *
+	 * @author Oliver Gierke
+	 */
+	private enum GuavaOptionalUnwrapper implements Converter<Object, Object> {
+
+		INSTANCE;
+
+		/*
+		 * (non-Javadoc)
+		 * @see org.springframework.core.convert.converter.Converter#convert(java.lang.Object)
+		 */
+		@Nullable
+		@Override
+		public Object convert(Object source) {
+			return source instanceof Optional ? ((Optional<?>) source).orNull() : source;
+		}
+	}
+
+	/**
+	 * A {@link Converter} to unwrap JDK 8 {@link java.util.Optional} instances.
+	 *
+	 * @author Oliver Gierke
+	 */
+	private enum Jdk8OptionalUnwrapper implements Converter<Object, Object> {
+
+		INSTANCE;
+
+		/*
+		 * (non-Javadoc)
+		 * @see org.springframework.core.convert.converter.Converter#convert(java.lang.Object)
+		 */
+		@Nullable
+		@Override
+		public Object convert(Object source) {
+			return source instanceof java.util.Optional ? ((java.util.Optional<?>) source).orElse(null) : source;
+		}
+	}
+
+	/**
+	 * A {@link Converter} to unwrap a Scala {@link Option} instance.
+	 *
+	 * @author Oliver Gierke
+	 * @author Mark Paluch
+	 * @since 1.12
+	 */
+	private enum ScalOptionUnwrapper implements Converter<Object, Object> {
+
+		INSTANCE;
+
+		private final Function0<Object> alternative = new AbstractFunction0<Object>() {
+
+			/*
+			 * (non-Javadoc)
+			 * @see scala.Function0#apply()
+			 */
+			@Nullable
+			@Override
+			public Option<Object> apply() {
+				return null;
+			}
+		};
+
+		/*
+		 * (non-Javadoc)
+		 * @see org.springframework.core.convert.converter.Converter#convert(java.lang.Object)
+		 */
+		@Nullable
+		@Override
+		public Object convert(Object source) {
+			return source instanceof Option ? ((Option<?>) source).getOrElse(alternative) : source;
+		}
+	}
+
+	/**
+	 * Converter to unwrap Vavr {@link io.vavr.control.Option} instances.
+	 *
+	 * @author Oliver Gierke
+	 * @since 2.0
+	 */
+	private enum VavrOptionUnwrapper implements Converter<Object, Object> {
+
+		INSTANCE;
+
+		/*
+		 * (non-Javadoc)
+		 * @see org.springframework.core.convert.converter.Converter#convert(java.lang.Object)
+		 */
+		@Nullable
+		@Override
+		@SuppressWarnings("unchecked")
+		public Object convert(Object source) {
+
+			if (source instanceof io.vavr.control.Option) {
+				return ((io.vavr.control.Option<Object>) source).getOrElse(() -> null);
+			}
+
+			return source;
+		}
+	}
+
+	public static final class WrapperType {
```

Comment:
This is probably an exact copy of `QueryExecutionConverters.WrapperType`. Shouldn't this be refactored to use the same class? Or is that not possible for backwards-compatibility reasons?

Reply:
Ok, clear! Looking forward to using this in a future release :-)